### PR TITLE
tests work

### PR DESCRIPTION
**Pull request by kzangeli**
Unit tests and func tests now work with the new sub-cache implementation.
The functests implemented for 1308 included in this PR - I still haven't seen what exactly they do.
Next PR, the last one for this issue, will be functests especially tailored for the new subcache.

### CODE REVIEW

File: src/lib/mongoBackend/TriggeredSubscription.cpp [hunk 1]
```diff
@@ -35,13 +35,13 @@
...
-  AttributeList      _attrL,
-  const std::string& _cacheSubId,
-  const std::string& _tenant
+  long long           _throttling,
+  long long           _lastNotification,
+  Format              _format,
+  const std::string&  _reference,
+  AttributeList       _attrL,
+  const std::string&  _cacheSubId,
+  const char*         _tenant
```
Comment (fgalan) [inline src/lib/mongoBackend/TriggeredSubscription.cpp]: Taking into account that the other paramters (_reference and _cacheSubId) are strings, shouldn't be this also a string? Or maybe there is a strong reason to change it to char*...

File: test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test [hunk 2]
```diff
@@ -56,7 +56,6 @@ Date: REGEX(.*)
   <dbConnectionPoolWaitingTime>REGEX(\d+.\d+)</dbConnectionPoolWaitingTime>
   <transactionSemaphoreWaitingTime>REGEX(\d+.\d+)</transactionSemaphoreWaitingTime>
   <curlContextMutexWaitingTime>REGEX(\d+.\d+)</curlContextMutexWaitingTime>
-  <subCacheMutexWaitingTime>REGEX(\d+.\d+)</subCacheMutexWaitingTime>
```
Comment (fgalan) [inline test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test]: This is going to be recovered in a future PR?

File: test/functionalTest/cases/0563_ngsi10_get_list_of_all_entities/all_entities_no_filter.test [hunk 3]
```diff
@@ -320,7 +320,6 @@ Date: REGEX(.*)
   <allContextEntitiesRequests>2</allContextEntitiesRequests>
   <versionRequests>0</versionRequests>
   <statisticsRequests>1</statisticsRequests>
-  <subCacheLookups>3</subCacheLookups>
```
Comment (fgalan) [inline test/functionalTest/cases/0563_ngsi10_get_list_of_all_entities/all_entities_no_filter.test]: This is going to be recovered in a future PR?

File: test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test [hunk 4]
```diff
@@ -337,9 +333,6 @@ Date: REGEX(.*)
         "measuring_interval_in_secs": "REGEX(\d+)", 
         "requestSemaphoreWaitingTime": "REGEX(\d+.\d+)", 
         "statisticsRequests": "3", 
-        "subCacheEntries": "REGEX((1|2))", 
-        "subCacheMutexWaitingTime": "REGEX(\d+.\d+)", 
-        "subCacheRemovals": "1",
```
Comment (fgalan) [inline test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test]: Same here... at the end, what should be recovered (or not) are:

* subCacheEntries
* subCacheMutexWaitingTime
* subCacheRemovals

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Possible Bug in Subscription Cache when reloading from database
+
+--SHELL-INIT--
+dbInit CB
+dbInit ftest-t1
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: Environment variables should be used for ftest-t1 and ftest-t2.

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Possible Bug in Subscription Cache reloading from database for 0.24.0
+
+--SHELL-INIT--
+dbInit CB
+dbInit ftest-t1
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test]: Same here.

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+  ]
+}'
+orionCurl --url /v1/subscribeContext --payload "$payload" --json --tenant t2
+echo
+echo
+
+
+echo "04. See subscriptions w/o tenant"
+echo "================================"
+echo 'db.csubs.findOne()' | mongo ftest
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test]: mongoCmd() should be used (maybe adapting it, if need). Also for similar steps in this .test.

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Temp
+
+--SHELL-INIT--
+dbInit CB
+dbInit ftest-t1
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Same here

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Temp
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Temporal name should be changed.

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: /v2/entities/s01
+Date: REGEX(.*)
+
+
+
+--TEARDOWN--
+brokerStop CB
+#dbDrop CB
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Uncomment (review it in other .tests, pleas).

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+brokerStart CB 216 IPv4 -multiservice -subCacheIval 2
+
+--SHELL--
+
+#
+# 01. Create subscription for s0.* entities, attribute A, w/o tenant
+# 02. Create subscription for s1.* entities, attribute A, for tenant t1
+# 03. Wait 3 secs so that the subscription cache has been reset from DB
+# 04. Create entity s01 w/ attr A w/o tenant
+#
```
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Maybe an accumulator dump step is missing at 5?

File: test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp [hunk 12]
```diff
@@ -280,19 +280,23 @@ static void prepareDatabase(bool initializeCache = true) {
...
   connection->insert(SUBSCRIBECONTEXT_COLL, sub1);
   connection->insert(SUBSCRIBECONTEXT_COLL, sub2);
   connection->insert(SUBSCRIBECONTEXT_COLL, sub3);
 
   /* Given that preparation including csubs, we have to init cache */
-  if (initializeCache == true)
+  if (useSubCache == true)
   {
     mongoSubCacheInit();
+    mongoSubCacheRefresh();
```
Comment (fgalan) [inline test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp]: I understand that now mongoSubCacheInit() doesn't include the initial refresh, isn't it?

File: test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test]: There is no longer any mutex for the subcache ... :-)

File: src/lib/mongoBackend/TriggeredSubscription.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/mongoBackend/TriggeredSubscription.cpp]: Yes, the tenant can be NULL (no tenant) in the new sub-cache. If NULL is used with const std::string&, it doesn't work

File: test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test]: Yes, I already have these counters back, coming in my next PR, with specific tests

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: What?
CB is also not an env var.
There are NO env vars as input to dbInit. 
We have never had that.
And frankly, what on earth would we gain by hiding the string "ftest-t1" in a variable?

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 11 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: This test WAS temporal, I added it by mistake.
The mistake is already fixed - coming in the next PR

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 9 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: This test WAS temporal, I added it by mistake.
The mistake is already fixed - coming in the next PR

File: src/lib/mongoBackend/TriggeredSubscription.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/mongoBackend/TriggeredSubscription.cpp]: Ok, that's a very powerfull reason :)

NTC

File: test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/1048_subscription_cache/three_subscriptions.test]: Ok, let's check it in the next PR.

NTC

File: test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp (hunk 12 above)
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp]: Yeah, now the init function does only what a constructor would do.
I am not sure this is the way to go.
Perhaps mongoSubCacheInit should call mongoSubCacheRefresh ...

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: However, you are using a name that depends of a env var CB_DATABASE_NAME (ftest), so if somebody changes CB_DATABASE_NAME this instruction breaks.

Maybe the good solution would be to have a function to initialize tenants. Something like:

dbInitTenant CB f1

However, not for this PR. I will open an issue about it to address it in the future.

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 9 above)
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Ok, let's check it in the next PR. NTC.

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: OK, now I understand you (I thought you had gone crazy! :-D)
If the base-name of the 'tenants' is 'xxx-' and this DB is called 'yyy-tenX', then its tenant will be '' (empty).
Instead og givern the *entire* name of the DB, only the 'tenant-path' should be given.
I am sure we have this problem in a number of places. An issue will be necessary.

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 11 above)
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Ok, let's check in the next PR.

File: test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp (hunk 12 above)
Comment (fgalan) [inline test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp]: NTC

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: What we should do is probably to change dbInit (and dbDrop) to accept a tenant name as input and not a database name.

Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db.test]: Issue https://github.com/telefonicaid/fiware-orion/issues/1374

NTC

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test]: Issue https://github.com/telefonicaid/fiware-orion/issues/1374

NTC

File: test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/0000_statistics_operation/statistics_with_semaphores.test]: NTC (discussed at skype)

File: test/functionalTest/cases/0563_ngsi10_get_list_of_all_entities/all_entities_no_filter.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/0563_ngsi10_get_list_of_all_entities/all_entities_no_filter.test]: NTC

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 8 above)
Comment (fgalan) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: Issue #1374

NTC

File: test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp (hunk 12 above)
Comment (kzangeli) [inline test/unittests/mongoBackend/mongoUpdateContext_withOnchangeSubscriptions_test.cpp]: Wait, let's discuss this ...
mongoSubCacheRefresh or not in mongoSubCacheInit?
I don't have a strong opinion on this ...
Let's do it on skype

File: test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_reset_from_db0_24_0.test]: This test has been removed, it doesn't make any sense to check for what's in mongo when testing the sub-cache. 
But, removed in the next PR, the one where I actually spent some time looking at these tests for 1308

File: test/functionalTest/cases/1308_subscription_cache/cache_traces.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/1308_subscription_cache/cache_traces.test]: I checked, this line is commented only in cache_traces.test.
And cache_traces.test is removed in the coming PR so ...

NTC

**Comment by fgalan** [comment]
LGTM

**Pull request merged**